### PR TITLE
Add C APIs and sample C application for existing GO Library.

### PR DESCRIPTION
**Pull request by Billy99**
Signed-off-by: Billy McFall <22157057+Billy99@users.noreply.github.com>

### CODE REVIEW

File: samples/c_app/app_sample.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+				free(networkStatusRsp.Status[i].IPs[j]);
+			}
+		}
+
+		if (networkStatusRsp.Status[i].Mac) {
+			printf("  networkStatusRsp.Status[%d].Mac = %s\n", i, networkStatusRsp.Status[i].Mac);
+			free(networkStatusRsp.Status[i].Mac);
+		}
+	}
+}
```
Comment (zenghui-shi) [inline samples/c_app/app_sample.c:102]: nit: build warning at the end of `main`. no `return`.

File: c_api/netutil_c_api.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#define NETUTIL_NUM_NETWORKSTATUS  10
+
+struct NetworkStatus {
+    char*    Name;
+    char*    Interface;
+    char*    IPs[NETUTIL_NUM_IPS];
+    char*    Mac;
+};
+struct NetworkStatusResponse {
+	struct NetworkStatus Status[NETUTIL_NUM_NETWORKSTATUS];
```
Comment (zenghui-shi) [inline c_api/netutil_c_api.go:34]: This looks like a fixed array length, I assume C doesn't support variable length array,  if the C code cannot predict how much memory needs be allocated to hold all the data, is there any other way to let the C code know how much memory shall be allocated?

File: Makefile [hunk 3]
```diff
@@ -2,6 +2,9 @@ default:
 	./hack/build.sh
 image:
 	./hack/build-image.sh
+c_sample:
```
Comment (zenghui-shi) [inline Makefile:5]: what about moving this `c_sample` build cmds into `hack/build.sh`, where it allows building outside system `GOPATH`.

File: samples/c_app/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+4. Test
+To set, run the application binary:
+```
+$ ./bin/c_sample
+```
+
+If the application is not actually running in a container where annotations have been
+exposed, run the following to copy a sample annotation file onto the system:
+```
+$ sudo cp samples/c_app/annotations /etc/poding/.
```
Comment (zenghui-shi) [inline samples/c_app/README.md]: Very clear doc! one minor comment:
`/etc/poding` usually does not exist, perhaps adding a cmd `mkdir -p /etc/poding`
nit: `/etc/podnetinfo`  is now used in admission controller, might change the file path `/etc/poding`.

File: Makefile (hunk 3 above)
Comment (Billy99) [reply Makefile:5]: I started down that path and ran into some issue (laptop rebooted so don't have the error in my history), so went down this path to make progress on the code. No problem with what you are suggesting, just couldn't figure out how to make it work (but didn't spend much time on it).

File: samples/c_app/README.md (hunk 4 above)
Comment (Billy99) [reply samples/c_app/README.md]: Typo in my documentation. I am using /etc/podinfo/. I also knew Admission Controller was using /etc/podnetinfo/, but the const netStatusPath in network.go in master is still using /etc/podinfo/.

File: c_api/netutil_c_api.go (hunk 2 above)
Comment (Billy99) [reply c_api/netutil_c_api.go:34]: Yes, need to figure this out. C doesn't have the dynamic nature with arrays. Only issue with calling to get the size is a race condition where the length could change (not likely) between calls. I'll put more thought into it.

Also, this PR just mapped the existing APIs. In the end, need more useful APIs, like get all interfaces, not get raw data (i.e. all environmental variables).

File: samples/c_app/app_sample.c (hunk 1 above)
Comment (Billy99) [reply samples/c_app/app_sample.c:102]: Ack

File: c_api/netutil_c_api.go (hunk 2 above)
Comment (zenghui-shi) [reply c_api/netutil_c_api.go:34]: I think it's fine to have the fixed array length until we find a better way to determine the array size.
Agree that this is not the final version of APIs we want to deliver to user, it's more of a sample showing how the API works.
we should be good with current version of PR.

File: samples/c_app/README.md (hunk 4 above)
Comment (zenghui-shi) [reply samples/c_app/README.md]: let's use the same `/etc/podinfo/` path as defined in const.
we can change the two later with a separate PR.

**Comment by zenghui-shi** [comment]
/lgtm
Thanks for adding notes in `app_sample.c` code!

**Pull request merged**